File: Backend/mapping/src/main/java/com/mhp/coding/challenges/mapping/mappers/ArticleMapper.java
package com.mhp.coding.challenges.mapping.mappers;

import com.mhp.coding.challenges.mapping.models.db.Article;
import com.mhp.coding.challenges.mapping.models.db.blocks.*;
import com.mhp.coding.challenges.mapping.models.dto.ArticleDto;
import com.mhp.coding.challenges.mapping.models.dto.blocks.*;
import org.mapstruct.AfterMapping;
import org.mapstruct.Mapper;
import org.mapstruct.MappingTarget;
import org.mapstruct.ReportingPolicy;

import java.util.stream.Collectors;

@Mapper(
        componentModel = "spring",
        unmappedTargetPolicy = ReportingPolicy.ERROR
)
public interface ArticleMapper {

    ArticleDto articleToArticleDto(Article article);

    default ArticleBlockDto articleBlockToArticleBlockDto(ArticleBlock articleBlock) {
        if (articleBlock instanceof GalleryBlock) {
            return galleryBlockToGalleryBlockDto((GalleryBlock) articleBlock);
        } else if (articleBlock instanceof ImageBlock) {
            return imageBlockToImageBlockDto((ImageBlock) articleBlock);
        } else if (articleBlock instanceof TextBlock) {
            return textBlockToTextBlockDto((TextBlock) articleBlock);
        } else if (articleBlock instanceof VideoBlock) {
            return videoBlockToVideoBlockDto((VideoBlock) articleBlock);
        } else {
            throw new UnsupportedOperationException("Mapping is not implemented for ArticleBlocks of type " + articleBlock.getClass());
        }
    }

    GalleryBlockDto galleryBlockToGalleryBlockDto(GalleryBlock galleryBlock);

    ImageBlockDto imageBlockToImageBlockDto(ImageBlock imageBlock);

    TextBlockDto textBlockToTextBlockDto(TextBlock textBlock);

    VideoBlockDto videoBlockToVideoBlockDto(VideoBlock videoBlock);

    @AfterMapping
    default void sortBlocksAfterArticleMapping(@MappingTarget ArticleDto articleDto) {
        articleDto.setBlocks(articleDto.getBlocks().stream()
                .sorted()
                .collect(Collectors.toList()));
    }

    default Article map(ArticleDto articleDto) {
        // Nicht Teil dieser Challenge.
        return new Article();
    }
}

File: Backend/mapping-kotlin/src/main/kotlin/com/mhp/coding/challenges/mapping/mappers/ArticleMapper.kt
package com.mhp.coding.challenges.mapping.mappers

import com.mhp.coding.challenges.mapping.models.db.Article
import com.mhp.coding.challenges.mapping.models.db.Image
import com.mhp.coding.challenges.mapping.models.db.ImageSize
import com.mhp.coding.challenges.mapping.models.db.blocks.*
import com.mhp.coding.challenges.mapping.models.dto.ArticleDto
import com.mhp.coding.challenges.mapping.models.dto.ImageDto
import com.mhp.coding.challenges.mapping.models.dto.blocks.*
import org.springframework.stereotype.Component
import java.util.*

@Component
class ArticleMapper {

    fun mapToDto(article: Article): ArticleDto {
        return ArticleDto(
            id = article.id,
            title = article.title,
            description = article.description.orEmpty(),
            author = article.author.orEmpty(),
            blocks = article.blocks
                .map(this::mapToDto)
                .sortedBy(ArticleBlockDto::sortIndex)
        )
    }

    fun mapToDto(articleBlock: ArticleBlock): ArticleBlockDto {
        return when (articleBlock) {
            is GalleryBlock -> mapToDto(articleBlock)
            is ImageBlock -> mapToDto(articleBlock)
            is TextBlock -> mapToDto(articleBlock)
            is VideoBlock -> mapToDto(articleBlock)
            else -> {
                throw UnsupportedOperationException("Mapping of class %s is not supported.".format(articleBlock.javaClass.simpleName))
            }
        }
    }

    fun mapToDto(galleryBlock: GalleryBlock): GalleryBlockDto {
        return GalleryBlockDto(
            images = galleryBlock.images
                .filterNotNull()
                .map(this::mapToDto),
            sortIndex = galleryBlock.sortIndex
        )
    }

    fun mapToDto(image: Image?): ImageDto {
        return ImageDto(
            id = image?.id ?: -1,
            url = image?.url.orEmpty(),
            imageSize = image?.imageSize ?: ImageSize.SMALL,
        )
    }

    fun mapToDto(imageBlock: ImageBlock): ImageBlockDto {
        return ImageBlockDto(
            image = mapToDto(imageBlock.image),
            sortIndex = imageBlock.sortIndex
        )
    }

    fun mapToDto(textBlock: TextBlock): TextBlockDto {
        return TextBlockDto(
            text = textBlock.text,
            sortIndex = textBlock.sortIndex
        )
    }

    fun mapToDto(videoBlock: VideoBlock): VideoBlockDto {
        return VideoBlockDto(
            url = videoBlock.url,
            type = videoBlock.type,
            sortIndex = videoBlock.sortIndex
        )
    }

    // Not part of the challenge / Nicht Teil dieser Challenge.
    fun map(articleDto: ArticleDto?): Article = Article(
        title = "An Article",
        blocks = emptySet(),
        id = 1,
        lastModified = Date()
    )
}
